docs/source/extensions/callbacks.rst

```diff
@@ -119,10 +119,61 @@ Persisting State
 Some callbacks require internal state in order to function properly. You can optionally
 choose to persist your callback's state as part of model checkpoint files using the callback hooks
 :meth:`~pytorch_lightning.callbacks.Callback.on_save_checkpoint` and :meth:`~pytorch_lightning.callbacks.Callback.on_load_checkpoint`.
-However, you must follow two constraints:
+Note that the returned state must be able to be pickled.
 
-1. Your returned state must be able to be pickled.
-2. You can only use one instance of that class in the Trainer callbacks list. We don't support persisting state for multiple callbacks of the same class.
+When your callback is meant to be used only as a singleton callback then implementing the above two hooks is enough
+to persist state effectively. However, if passing multiple instances of the callback to the Trainer is supported, then
+the callback must define a :attr:`~pytorch_lightning.callbacks.Callback.state_id` property in order for Lightning
+to be able to distinguish the different states when loading the callback state. This concept is best illustrated by
+the following example.
+
+.. testcode::
+
+    class Counter(Callback):
+        def __init__(self, what="epochs", verbose=True):
+            self.what = what
+            self.verbose = verbose
+            self.state = {"epochs": 0, "batches": 0}
+
+        @property
+        def state_id(self):
+            # note: we do not include `verbose` here on purpose
+            return self._generate_state_id(what=self.what)
```

Comment:
Should this function be public if we advertise it?

Reply:
Maybe not make it public as a public method, because everything in the callback so far is a hook but this one here should not be overridden. 

Maybe just convert it to a public helper function

Reply:
>  because everything in the callback so far is a hook but this one here should not be overridden.

What would be the issue? All the hooks start with `on_*` so it should be easy to distinguish

Reply:
but we do not want the user to override "generate_state_id" when they create a callback. We want them to override the state_id property if needed.

Reply:
This right now is the only place where we use it. It's located where the documentation is and it's convenient to access. To move it, we would need a better motivation imo. For example, a valid motivation would be that we want to have other objects own a state id and we need to share code.